primitives/arithmetic/src/traits.rs

```diff
@@ -302,3 +307,519 @@ pub trait SaturatedConversion {
+		///     1.ensure_div(0)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     FixedI64::from(i64::MIN).ensure_div(FixedI64::from(-1))?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(extrinsic_zero(), Err(ArithmeticError::DivisionByZero.into()));
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// ```
+		fn ensure_div(self, v: Self) -> Result<Self, ArithmeticError> {
+			self.checked_div(&v).ok_or_else(|| error::division(self, v))
+		}
+	}
+
+	impl<T: CheckedAdd + PartialOrd + Zero + Copy> EnsureAdd for T {}
+	impl<T: CheckedSub + PartialOrd + Zero + Copy> EnsureSub for T {}
+	impl<T: CheckedMul + PartialOrd + Zero + Copy> EnsureMul for T {}
+	impl<T: CheckedDiv + PartialOrd + Zero + Copy> EnsureDiv for T {}
+
+	/// Meta trait that supports all inmutable arithmetic operations
+	pub trait EnsureOp: EnsureAdd + EnsureSub + EnsureMul + EnsureDiv {}
+	impl<T: EnsureAdd + EnsureSub + EnsureMul + EnsureDiv> EnsureOp for T {}
+
+	/// Performs self addition that returns `ArithmeticError` instead of wrapping around on
+	/// overflow.
+	pub trait EnsureAddAssign: EnsureAdd {
+		/// Adds two numbers overwriting the left hand one, checking for overflow.
+		/// If overflow happens, `ArithmeticError` is returned.
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureAddAssign, ArithmeticError};
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     let mut max = u32::MAX;
+		///     max.ensure_add_assign(1)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     let mut max = i32::MIN;
+		///     max.ensure_add_assign(-1)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_add_assign(&mut self, v: Self) -> Result<&mut Self, ArithmeticError> {
+			*self = self.ensure_add(v)?;
+			Ok(self)
+		}
+	}
+
+	/// Performs self subtraction that returns `ArithmeticError` instead of wrapping around on
+	/// underflow.
+	pub trait EnsureSubAssign: EnsureSub {
+		/// Subtracts two numbers overwriting the left hand one, checking for overflow.
+		/// If overflow happens, `ArithmeticError` is returned.
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureSubAssign, ArithmeticError};
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     let mut zero: u32 = 0;
+		///     zero.ensure_sub_assign(1)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     let mut zero = i32::MAX;
+		///     zero.ensure_sub_assign(-1)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// ```
+		fn ensure_sub_assign(&mut self, v: Self) -> Result<&mut Self, ArithmeticError> {
```

Comment:
Not sure what is best practice about returning here, I think on other modification functions we return `()`. @bkchr ?

Reply:
The returned value is useful when you want to chain some operations:
```rust
value
    .ensure_add_assign(23)?
    .ensure_mul_assign(other)?
    .ensure_sub_assign(42)?;
```

But it's true that the normal `OpsAssign` methods do not use the return value, so I can leave the with `()`.

Reply:
I was re-thinking this, And it seems like returning `()` could be a better option.

Returning a value could be less ergonomic for cases where the method ends with that line, where you're forced to write an extra `Ok(())`:

```rust
fn foo() -> Result<(), ArithmeticError> {
    value.ensure_add_assign(23)?;
    Ok(()) // Ouch! I need to write this
}
```

And the first example can be rewritten as follows:

```rust
value.ensure_add_assign(23)?
value.ensure_mul_assign(other)?
value.ensure_sub_assign(42)?;
```

What do you think?

Reply:
@ggwpez @melekes Finally, I've changed this to return `()` because is the natural expected behavior. If you do not agree, I can revert the last commit.